### PR TITLE
Add IM CommandPathParams/AttributePathParams/EventPathParams

### PR DESCRIPTION
**Pull request by yunhanw-google**
-- Move/Rename CommandPathParams in IM utilities
-- Add AttributePathParams, EventPathParams, ClusterInfo to IM utilities header
so that other IM can directly include this and avoid forward
declarations and header include loop.

### CODE REVIEW

File: src/app/InteractionModelUtilities.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/*
```
Comment (andy31415) [inline src/app/InteractionModelUtilities.h:1]: could we avoid having names like utilites/helpers/support?

These seem to be 'structures' in general, but unsure if InteractionModelStructures gives me sufficient info of what this contains. Is there a more descriptive name? what do all these structures here have in common?

Or alternativel, if they do not have something specific in common, consider splitting into separate headesrs like `CommandPathParams.h`, `AttributePathParams.h` etc.

Or `InteractionModelPathParams.g` and `ClusterInfo.h`

File: src/app/InteractionModelUtilities.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    CommandPathParams(EndpointId aEndpointId, GroupId aGroupId, ClusterId aClusterId, CommandId aCommandId,
+                      const BitFlags<CommandPathFlags> & aFlags) :
+        mEndpointId(aEndpointId),
+        mGroupId(aGroupId), mClusterId(aClusterId), mCommandId(aCommandId), mFlags(aFlags)
+    {}
+    bool operator==(const CommandPathParams & aCommandPathParams) const
+    {
+        return ((aCommandPathParams.mClusterId == mClusterId) && (aCommandPathParams.mCommandId == mCommandId) &&
+                (mFlags.Has(aCommandPathParams.mFlags)) &&
+                (mFlags.Has(CommandPathFlags::kEndpointIdValid) ? aCommandPathParams.mEndpointId == mEndpointId : false) &&
```
Comment (andy31415) [inline src/app/InteractionModelUtilities.h:44]: This only checks the flags in `this`. How about flags in the other one? What if mFlags.Has(kEndpointIdValid) and but `! aCommandPathParams.Has(kEndpointIdValid)` ?

Could you add unit tests for this?

File: src/app/InteractionModelUtilities.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                        ListIndex aListIndex, const BitFlags<AttributePathFlags> & aFlags) :
+        mNodeId(aNodeId),
+        mEndpointId(aEndpointId), mClusterId(aClusterId), mFieldId(aFieldId), mListIndex(aListIndex), mFlags(aFlags)
+    {}
+
+    bool operator==(const AttributePathParams & aAttributePathParams) const
+    {
+        return ((aAttributePathParams.mNodeId == mNodeId) && (aAttributePathParams.mEndpointId == mEndpointId) &&
+                (aAttributePathParams.mClusterId == mClusterId) && (mFlags.Has(aAttributePathParams.mFlags)) &&
+                (mFlags.Has(AttributePathFlags::kFieldIdValid) ? aAttributePathParams.mFieldId == mFieldId : false) &&
```
Comment (andy31415) [inline src/app/InteractionModelUtilities.h:72]: same here - it looks as if we should require mFlags to be identical

File: src/app/InteractionModelUtilities.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+};
+
+struct ClusterInfo
+{
+    bool IsDirty(void) { return mDirty; }
+    void SetDirty(void) { mDirty = true; }
+    void ClearDirty(void) { mDirty = false; }
+    bool operator==(const ClusterInfo & aClusterInfo) const
+    {
+        return ((aClusterInfo.mAttributePathParams == mAttributePathParams));
```
Comment (andy31415) [inline src/app/InteractionModelUtilities.h:109]: why is dirty not used? a comment would be useful. Also remove double parantheses.

File: src/app/InteractionModelUtilities.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    NodeId mNodeId         = 0;
+    EndpointId mEndpointId = 0;
+    ClusterId mClusterId   = 0;
+    EventId mEventId       = 0;
+    bool mIsUrgent         = false;
+};
+
+struct ClusterInfo
+{
+    bool IsDirty(void) { return mDirty; }
```
Comment (andy31415) [inline src/app/InteractionModelUtilities.h:104]: nit: remove `(void)` as argument - CHIP style is to not have this pattern (mainly because there are clang cleanups that can remove it but not the other way around)

File: src/app/InteractionModelUtilities.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/InteractionModelUtilities.h:72]: (mFlags.Has(aAttributePathParams.mFlags) which would do same flag check between this and aAttributePathParams

File: src/app/InteractionModelUtilities.h (hunk 4 above)
Comment (yunhanw-google) [reply src/app/InteractionModelUtilities.h:109]: replace operator with IsSamePath, which mean only check if path is same

**Review by msandstedt — APPROVED** [review]

File: src/app/InteractionModelUtilities.h (hunk 1 above)
Comment (yunhanw-google) [reply src/app/InteractionModelUtilities.h]: Done, thanks

File: src/app/InteractionModelUtilities.h (hunk 2 above)
Comment (yunhanw-google) [reply src/app/InteractionModelUtilities.h]: Done, thanks

File: src/app/InteractionModelUtilities.h (hunk 5 above)
Comment (yunhanw-google) [reply src/app/InteractionModelUtilities.h]: Done, thanks

File: src/app/AttributePathParams.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+enum class AttributePathFlags : uint8_t
+{
+    kFieldIdValid   = 0x01,
+    kListIndexValid = 0x02,
+};
+
+struct AttributePathParams
+{
+    AttributePathParams(NodeId aNodeId, EndpointId aEndpointId, ClusterId aClusterId, FieldId aFieldId, ListIndex aListIndex,
+                        const BitFlags<AttributePathFlags> & aFlags) :
```
Comment (Damian-Nordic) [inline src/app/AttributePathParams.h:35]: aFlags is of size 1B, so it can be passed by-value as well.

File: src/app/AttributePathParams.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        mClusterId  = aAttributePathParams.mClusterId;
+        mFieldId    = aAttributePathParams.mFieldId;
+        mListIndex  = aAttributePathParams.mListIndex;
+        mFlags      = aAttributePathParams.mFlags;
+    }
+    bool IsSamePath(const AttributePathParams & aAttributePathParams) const
+    {
+        return aAttributePathParams.mNodeId == mNodeId && aAttributePathParams.mEndpointId == mEndpointId &&
+                aAttributePathParams.mClusterId == mClusterId && mFlags.Has(aAttributePathParams.mFlags) &&
+                mFlags.Has(AttributePathFlags::kFieldIdValid)
```
Comment (Damian-Nordic) [inline src/app/AttributePathParams.h:52]: This doesn't look correct. How is this supposed to work? If you use `mFlags.Has`, which only checks if there's an intersection between both bitmasks, then you probably should check if `AttributePathFlags::kFieldIdValid` belongs to the intersection, not to `mFlags`, right? Also, the ternary operator has lower priority than `&&`, so you should probably add parentheses.

File: src/app/ClusterInfo.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+#pragma once
+
+#include <app/AttributePathParams.h>
+#include <app/util/basic-types.h>
+
+namespace chip {
+namespace app {
+struct ClusterInfo
+{
+    ClusterInfo(AttributePathParams & aAttributePathParams, bool aDirty) :
```
Comment (Damian-Nordic) [inline src/app/ClusterInfo.h:28]: `const AttributePathParams&` ?

File: src/app/CommandPathParams.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+{
+    CommandPathParams(EndpointId aEndpointId, GroupId aGroupId, ClusterId aClusterId, CommandId aCommandId,
+                      const BitFlags<CommandPathFlags> & aFlags) :
+        mEndpointId(aEndpointId),
+        mGroupId(aGroupId), mClusterId(aClusterId), mCommandId(aCommandId), mFlags(aFlags)
+    {}
+    bool IsSamePath(const CommandPathParams & aCommandPathParams) const
+    {
+        return aCommandPathParams.mClusterId == mClusterId && aCommandPathParams.mCommandId == mCommandId &&
+                mFlags.Has(aCommandPathParams.mFlags) && mFlags.Has(CommandPathFlags::kEndpointIdValid)
```
Comment (Damian-Nordic) [inline src/app/CommandPathParams.h:43]: Same comment as for AttributePathParams. The relation doesn't seem symmetric and probably parentheses should be added around the ternary operators.

File: src/app/tests/TestAttributePathParams.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+{
+    AttributePathParams attributePathParams1(1, 2, 3, 4, 5, AttributePathFlags::kListIndexValid);
+    AttributePathParams attributePathParams2(1, 2, 3, 4, 6, AttributePathFlags::kListIndexValid);
+    NL_TEST_ASSERT(apSuite, !attributePathParams1.IsSamePath(attributePathParams2));
+}
+
+void TestDifferentPathFlag(nlTestSuite * apSuite, void * apContext)
+{
+    AttributePathParams attributePathParams1(1, 2, 3, 4, 5, AttributePathFlags::kFieldIdValid);
+    AttributePathParams attributePathParams2(1, 2, 3, 4, 5, AttributePathFlags::kListIndexValid);
```
Comment (Damian-Nordic) [inline src/app/tests/TestAttributePathParams.cpp:77]: What if `attributePathParams1` contains both `fieldId` and `listIndex` while `attributePathParams2` - only `listIndex`? Should we test such scenarios?

File: src/app/CommandPathParams.h (hunk 9 above)
Comment (andy31415) [reply src/app/CommandPathParams.h:43]: I would propose to not try to do everything in one return since this results in very hard to follow code.

Use:

```
  if (mFlags != aCommandPathParams.mFlags) {
     return false;
  }

  if (mFlags.Has(CommandPathFlags::kEndpointIdValid)) && aCommandPathParams.mEndpointId != mEndpointId) {
     return false;
  }
  /// ...
```

Then we do not have to worry about priorities of ternary nor do we have to parse a super long bool expression

File: src/app/CommandPathParams.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+struct CommandPathParams
+{
+    CommandPathParams(EndpointId aEndpointId, GroupId aGroupId, ClusterId aClusterId, CommandId aCommandId,
+                      const BitFlags<CommandPathFlags> & aFlags) :
+        mEndpointId(aEndpointId),
+        mGroupId(aGroupId), mClusterId(aClusterId), mCommandId(aCommandId), mFlags(aFlags)
+    {}
+    bool IsSamePath(const CommandPathParams & aCommandPathParams) const
```
Comment (andy31415) [inline src/app/CommandPathParams.h:40]: nit: consider something like `other` as a name instead of `aCommandPathparams`. Shorter name seems to be easier to follow. Here and above as well.

File: src/app/CommandPathParams.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        return aCommandPathParams.mClusterId == mClusterId && aCommandPathParams.mCommandId == mCommandId &&
+                mFlags.Has(aCommandPathParams.mFlags) && mFlags.Has(CommandPathFlags::kEndpointIdValid)
+            ? aCommandPathParams.mEndpointId == mEndpointId
+            : false || mFlags.Has(CommandPathFlags::kGroupIdValid) ? aCommandPathParams.mGroupId == mGroupId : false;
+    }
+    EndpointId mEndpointId            = 0;
+    GroupId mGroupId                  = 0;
+    ClusterId mClusterId              = 0;
+    CommandId mCommandId              = 0;
+    BitFlags<CommandPathFlags> mFlags = CommandPathFlags::kEndpointIdValid;
```
Comment (andy31415) [inline src/app/CommandPathParams.h:51]: How about initializing this to 0? EndpointId is not yet valid.

File: src/app/tests/TestAttributePathParams.cpp (hunk 10 above)
Comment (yunhanw-google) [reply src/app/tests/TestAttributePathParams.cpp:77]: if you put AttributePathParams's mFlag with 0x03, it cannot be compiled at all, it can only accept AttributePathFlags::kFieldIdValid and  AttributePathFlags::kListIndexValid

File: src/app/CommandPathParams.h (hunk 11 above)
Comment (yunhanw-google) [reply src/app/CommandPathParams.h:40]: Done, thanks

File: src/app/CommandPathParams.h (hunk 12 above)
Comment (yunhanw-google) [reply src/app/CommandPathParams.h:51]: mFlags can only be initialized with CommandPathFlags::kEndpointIdValid or CommandPathFlags::kGroupIdValid, if you put 0, complication cannot pass

File: src/app/ClusterInfo.h (hunk 8 above)
Comment (yunhanw-google) [reply src/app/ClusterInfo.h:28]: AttributePathParams could be changed later.

File: src/app/CommandPathParams.h (hunk 9 above)
Comment (yunhanw-google) [reply src/app/CommandPathParams.h]: Done， thanks

File: src/app/AttributePathParams.h (hunk 7 above)
Comment (yunhanw-google) [reply src/app/AttributePathParams.h]: refactored. thanks

File: src/app/CommandPathParams.h (hunk 12 above)
Comment (yunhanw-google) [reply src/app/CommandPathParams.h]: just remove intiializer, it is default as 0

File: src/app/AttributePathParams.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        mListIndex  = aAttributePathParams.mListIndex;
+        mFlags      = aAttributePathParams.mFlags;
+    }
+    bool IsSamePath(const AttributePathParams & other) const
+    {
+        if (other.mNodeId != mNodeId || other.mEndpointId != mEndpointId || other.mClusterId != mClusterId)
+        {
+            return false;
+        }
+        if (!mFlags.Has(other.mFlags))
```
Comment (andy31415) [inline src/app/AttributePathParams.h:54]: could you add a comment on why this is not == but rather Has?

This makes the operation not commutative (i.e. if one has no flags and another has flags,  then we can have `a.IsSamePath(b)` but NOT `b.IsSamePath(a)`) which seems unexpected.

File: src/app/ClusterInfo.h (hunk 8 above)
Comment (andy31415) [reply src/app/ClusterInfo.h:28]: yes, but you are creating a copy here. This should be const.

File: src/app/CommandPathParams.h [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        mEndpointId(aEndpointId),
+        mGroupId(aGroupId), mClusterId(aClusterId), mCommandId(aCommandId), mFlags(aFlags)
+    {}
+    bool IsSamePath(const CommandPathParams & other) const
+    {
+        if (other.mClusterId != mClusterId || other.mCommandId != mCommandId)
+        {
+            return false;
+        }
+        if (!mFlags.Has(other.mFlags))
```
Comment (andy31415) [inline src/app/CommandPathParams.h:46]: same comment here: can or should we use == ?

Unsure about the implications? I could see a desire to have a partial path match a more complete path, however either this should commute or maybe naming should indicate that order matters.

Maybe `ContainsPathOf(...)` so logically I can state "1/" contains "1/2/3" but "1/2/3" does not contain "1"

File: src/app/ClusterInfo.h (hunk 8 above)
Comment (yunhanw-google) [reply src/app/ClusterInfo.h]: yes, done, thanks

File: src/app/CommandPathParams.h (hunk 14 above)
Comment (yunhanw-google) [reply src/app/CommandPathParams.h]: yep just use ==

File: src/app/AttributePathParams.h (hunk 13 above)
Comment (yunhanw-google) [reply src/app/AttributePathParams.h]: just use ==

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Comment by yunhanw-google** [comment]
@jelderton @hawk248 @bzbarsky-apple

File: src/app/InteractionModelDelegate.h [hunk 15]
```diff
@@ -111,8 +110,7 @@ class InteractionModelDelegate
     /**
      * Notification that a Command Send has received an Invoke Command Response and fails to process a command data element in that
      * command response
-     * @param[in]  apCommandSender A current command sender which can identify the command sender to the consumer, particularly
-     * during multiple command interactions
+     * @param[in]  apCommandSender A current command sender which can identify the command sender to the consumer, particularly w*
```
Comment (jelderton) [inline src/app/InteractionModelDelegate.h:113]: did this comment get chopped off by accident?

**Review by jelderton — APPROVED** [review]

File: src/app/InteractionModelDelegate.h (hunk 15 above)
Comment (yunhanw-google) [reply src/app/InteractionModelDelegate.h]: yep, accidental chopped, fixed, thanks
